[16.0][ADD] document_page_tag_print_control: restrict document printing based on tag configuration

BetoRaja opened the pull request:
@HaraldPanten @ValentinVinagre 
This module allows restricting document page printing based on assigned tags. A configurable is_not_printable field is added to document page tags to block printing when marked. Documents are considered non-printable if any of their tags is marked as such. The print button remains visible, but attempting to print blocked pages will trigger a warning. 

T-5241

Commit:
[ADD] document_page_tag_print_control: restrict document printing based on tag configuration

HaraldPanten reviewed (changes requested):
Some comments:

- Why did you design the boolean option as "is_printable"? By default, all pages are printable, so wouldn’t it make more sense to use a boolean like "is_not_printable" or "non_printable" instead? In other words, why do we need a "is_printable" flag (default true) instead of a "is_not_printable" one (default false)? This might simplify the module logic, don't you think?

- I tried the module, and it doesn't seem to be working. Here are the steps I followed:

1. Created "printable" and "non-printable" tags.
2. Set each tag's is_printable value to True/False.
3. Assigned the "non-printable" tag to a page.
4. I was still able to print the page selecting the "Document Page" option under the "Action" button.

My functional testing was made in Runboat.

If you make any design changes, please remember to update the README accordingly.

BetoRaja commented:
> Some comments:
> 
> * Why did you design the boolean option as "is_printable"? By default, all pages are printable, so wouldn’t it make more sense to use a boolean like "is_not_printable" or "non_printable" instead? In other words, why do we need a "is_printable" flag (default true) instead of a "is_not_printable" one (default false)? This might simplify the module logic, don't you think?
> * I tried the module, and it doesn't seem to be working. Here are the steps I followed:
> 
> 1. Created "printable" and "non-printable" tags.
> 2. Set each tag's is_printable value to True/False.
> 3. Assigned the "non-printable" tag to a page.
> 4. I was still able to print the page selecting the "Document Page" option under the "Action" button.
> 
> My functional testing was made in Runboat.
> 
> If you make any design changes, please remember to update the README accordingly.

@HaraldPanten Thanks for the feedback — you were absolutely right. We’ve updated the module to use is_not_printable instead of is_printable, and adjusted the views and translations accordingly.

Also, the module logic has changed: it now uses a warning system instead of hiding the print button, and this behavior works as expected. Additionally, users with the Manager role in Document Knowledge are now exempt from tag-based print restrictions.

HaraldPanten commented on document_print_control/readme/USAGE.md:
```diff
@@ -0,0 +1,11 @@
+To restrict the printing of document pages using tags:
+
+1. Go to Document Management → Configuration → Tags.
```
```suggestion
1. Go to Document Management > Configuration > Tags.
```

BetoRaja replied:
Changed

HaraldPanten commented on document_print_control/readme/USAGE.md:
```diff
@@ -0,0 +1,11 @@
+5. If you attempt to print such pages, the system will raise a warning and prevent the action.
+6. All tags allow printing by default for compatibility.
+
+Users with the **Manager** role in the "Document Knowledge" section (Settings → Users → Access Rights → Other) are not affected by this restriction. They can always print any document page.
```
```suggestion
Users with the **Manager** role in the "Document Knowledge" section (Settings > Users > Access Rights > Other) are not affected by this restriction. They can always print any document page.
```

BetoRaja replied:
Changed

HaraldPanten commented on document_print_control/readme/CONTRIBUTORS.md:
```diff
@@ -0,0 +1,5 @@
+
+> - [Sygel](https://www.sygel.es):
+>
+> > - Juan Alberto Raja Martinez\<<juan.raja@sygel.es>\>
+> > - Valentín Vinagre Urteaga\<<valentin.vinagre@sygel.es>\>
```
You should check this format. It's not correct

<img width="1538" height="614" alt="Captura de pantalla 2025-07-14 a las 15 30 21" src="https://github.com/user-attachments/assets/4cbc8fa2-b584-4958-bc77-71a2321246c5" />

BetoRaja replied:
Changed

HaraldPanten reviewed (changes requested):
Here are my comments.

An additional one: Shouldn't we call this module document_page_tag_print_control, instead?

BetoRaja commented:
> Here are my comments.
> 
> An additional one: Shouldn't we call this module document_page_tag_print_control, instead?

@HaraldPanten Yes,it makes more sense, I'll change it and all its references.

HaraldPanten reviewed (approved):
LGTM

ValentinVinagre commented on document_page_tag_print_control/__manifest__.py:
```diff
@@ -0,0 +1,21 @@
+    "category": "Knowledge",
+    "depends": [
+        "document_page",
+        "document_page_tag",
```
only 'document_page_tag' .

BetoRaja replied:
Done

ValentinVinagre commented on document_page_tag_print_control/models/document_page.py:
```diff
@@ -0,0 +1,30 @@
+                tag.is_not_printable for tag in record.tag_ids
+            )
+
+    def action_print_document_page(self):
```
please only 1 return.  change all function 👍🏻

BetoRaja replied:
Ok, i've done with only 1 return.

ValentinVinagre commented on document_page_tag_print_control/models/report_document_page.py:
```diff
@@ -0,0 +1,40 @@
+
+    @api.model
+    def _get_report_values(self, docids, data=None):
+        docs = self.env["document.page"].browse(docids)
```
use .exists()

ValentinVinagre commented on document_page_tag_print_control/models/report_document_page.py:
```diff
@@ -0,0 +1,40 @@
+        docs = self.env["document.page"].browse(docids)
+
+        if self.env.user.has_group("document_page.group_document_manager"):
+            return {
```
only 1 return please.

BetoRaja replied:
Ok, i've done with only 1 return

ValentinVinagre reviewed (approved):
LGTM 👍🏻

BetoRaja commented:
@etobella Could you please review this PR when you have a moment? It's ready for your feedback and potential merging.Thanks!!

etobella commented on document_page_tag_print_control/models/document_page.py:
```diff
@@ -0,0 +1,32 @@
+                tag.is_not_printable for tag in record.tag_ids
+            )
+
+    def action_print_document_page(self):
```
this function is not used... why are you creating it?

BetoRaja replied:
This function is needed because without it, the print restriction doesn’t work, since the standard report action bypasses these checks. It acts as the gatekeeper preventing unauthorized prints in our module.

Thanks a lot for your review! Let me know what you think.

etobella replied:
@BetoRaja Where do you execute this function in the views of users?

This action is only used on tests and this code. It is not present on document.page view.

Also, the code that makes the print function work is the code inside `report.document_page...`

BetoRaja replied:
It's true @etobella , sorry for the inconvenience, it was a residue of a previous version, which does not contribute anything, it has been functionally tested. Thank you very much and again sorry for the inconvenience. I'll delete it and modify the tests.

BetoRaja replied:
@etobella Could you please take a look , the changes are already.

etobella commented on document_page_tag_print_control/tests/test_document_page_tag_print_control.py:
```diff
@@ -0,0 +1,90 @@
+            {"name": "Non-Printable", "is_not_printable": True}
+        )
+        cls.doc_without_tags = cls.env["document.page"].create(
+            {"name": "Sin etiquetas", "content": "Contenido"}
```
Make the tests in english please 🙏 

I speak Spanish, but a lot of collaborators don't. It is nicer for them 😉

BetoRaja replied:
Okey, i have translated all the tests,thank you for your comment.

etobella reviewed (commented):
A small comment on the tests

etobella reviewed (approved):
/ocabot merge nobump

ValentinVinagre commented:
@BetoRaja port to v: 17 & 18 . thanks.
